File: typescript_client/src/operations.ts
/**
 * FastDFS Operations
 * 
 * This module implements all file operations (upload, download, delete, etc.)
 * for the FastDFS client.
 */

import {
  encodeHeader,
  decodeHeader,
  splitFileId,
  joinFileId,
  encodeMetadata,
  decodeMetadata,
  getFileExtName,
  readFileContent,
  writeFileContent,
  padString,
  unpadString,
  encodeInt64,
  decodeInt64,
  encodeInt32,
  decodeInt32,
} from './protocol';
import {
  FDFS_GROUP_NAME_MAX_LEN,
  FDFS_FILE_EXT_NAME_MAX_LEN,
  IP_ADDRESS_SIZE,
  TrackerCommand,
  StorageCommand,
  StorageServer,
  FileInfo,
  MetadataFlag,
  Metadata,
} from './types';
import {
  InvalidResponseError,
  NoStorageServerError,
  mapStatusToError,
} from './errors';
import { ConnectionPool } from './connection';

/**
 * Handles all FastDFS file operations
 * 
 * This class is used internally by the Client class.
 */
export class Operations {
  private trackerPool: ConnectionPool;
  private storagePool: ConnectionPool;
  private networkTimeout: number;
  private retryCount: number;

  constructor(
    trackerPool: ConnectionPool,
    storagePool: ConnectionPool,
    networkTimeout: number,
    retryCount: number
  ) {
    this.trackerPool = trackerPool;
    this.storagePool = storagePool;
    this.networkTimeout = networkTimeout;
    this.retryCount = retryCount;
  }

  /**
   * Uploads a file from the local filesystem
   */
  async uploadFile(
    localFilename: string,
    metadata?: Metadata,
    isAppender: boolean = false
  ): Promise<string> {
    const fileData = readFileContent(localFilename);
    const extName = getFileExtName(localFilename);
    return this.uploadBuffer(fileData, extName, metadata, isAppender);
  }

  /**
   * Uploads data from a buffer
   */
  async uploadBuffer(
    data: Buffer,
    fileExtName: string,
    metadata?: Metadata,
    isAppender: boolean = false
  ): Promise<string> {
    for (let attempt = 0; attempt < this.retryCount; attempt++) {
      try {
        return await this.uploadBufferInternal(data, fileExtName, metadata, isAppender);
      } catch (error) {
        if (attempt === this.retryCount - 1) {
          throw error;
        }
        await this.sleep((attempt + 1) * 1000);
      }
    }
    throw new Error('Upload failed after retries');
  }

  /**
   * Internal implementation of buffer upload
   */
  private async uploadBufferInternal(
    data: Buffer,
    fileExtName: string,
    metadata?: Metadata,
    isAppender: boolean = false
  ): Promise<string> {
    // Get storage server from tracker
    const storageServer = await this.getStorageServer('');

    // Get connection to storage server
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Prepare upload command
      const cmd = isAppender ? StorageCommand.UPLOAD_APPENDER_FILE : StorageCommand.UPLOAD_FILE;

      // Build request
      const extNameBytes = padString(fileExtName, FDFS_FILE_EXT_NAME_MAX_LEN);
      const storePathIndex = storageServer.storePathIndex;

      const bodyLen = 1 + FDFS_FILE_EXT_NAME_MAX_LEN + data.length;
      const reqHeader = encodeHeader(bodyLen, cmd, 0);

      // Send request
      await conn.send(reqHeader, this.networkTimeout);
      await conn.send(Buffer.from([storePathIndex]), this.networkTimeout);
      await conn.send(extNameBytes, this.networkTimeout);
      await conn.send(data, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      if (respHeader.length <= 0) {
        throw new InvalidResponseError('Empty response body');
      }

      const respBody = await conn.receiveFull(respHeader.length, this.networkTimeout);

      // Parse response
      if (respBody.length < FDFS_GROUP_NAME_MAX_LEN) {
        throw new InvalidResponseError('Response body too short');
      }

      const groupName = unpadString(respBody.slice(0, FDFS_GROUP_NAME_MAX_LEN));
      const remoteFilename = respBody.slice(FDFS_GROUP_NAME_MAX_LEN).toString('utf8');

      const fileId = joinFileId(groupName, remoteFilename);

      // Set metadata if provided
      if (metadata && Object.keys(metadata).length > 0) {
        try {
          await this.setMetadata(fileId, metadata, MetadataFlag.OVERWRITE);
        } catch {
          // Metadata setting failed, but file is uploaded
        }
      }

      return fileId;
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Gets a storage server from tracker for upload
   */
  private async getStorageServer(groupName: string): Promise<StorageServer> {
    const conn = await this.trackerPool.get();

    try {
      // Prepare request
      const cmd = groupName
        ? TrackerCommand.SERVICE_QUERY_STORE_WITH_GROUP_ONE
        : TrackerCommand.SERVICE_QUERY_STORE_WITHOUT_GROUP_ONE;
      const bodyLen = groupName ? FDFS_GROUP_NAME_MAX_LEN : 0;

      const header = encodeHeader(bodyLen, cmd, 0);
      await conn.send(header, this.networkTimeout);

      if (groupName) {
        const groupNameBytes = padString(groupName, FDFS_GROUP_NAME_MAX_LEN);
        await conn.send(groupNameBytes, this.networkTimeout);
      }

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      if (respHeader.length <= 0) {
        throw new NoStorageServerError();
      }

      const respBody = await conn.receiveFull(respHeader.length, this.networkTimeout);

      // Parse storage server info
      if (respBody.length < FDFS_GROUP_NAME_MAX_LEN + IP_ADDRESS_SIZE + 9) {
        throw new InvalidResponseError('Storage server response too short');
      }

      let offset = FDFS_GROUP_NAME_MAX_LEN;
      const ipAddr = unpadString(respBody.slice(offset, offset + IP_ADDRESS_SIZE));
      offset += IP_ADDRESS_SIZE;

      const port = decodeInt64(respBody.slice(offset, offset + 8));
      offset += 8;

      const storePathIndex = respBody.readUInt8(offset);

      return {
        ipAddr,
        port,
        storePathIndex,
      };
    } finally {
      this.trackerPool.put(conn);
    }
  }

  /**
   * Downloads a file from FastDFS
   */
  async downloadFile(fileId: string, offset: number = 0, length: number = 0): Promise<Buffer> {
    for (let attempt = 0; attempt < this.retryCount; attempt++) {
      try {
        return await this.downloadFileInternal(fileId, offset, length);
      } catch (error) {
        if (attempt === this.retryCount - 1) {
          throw error;
        }
        await this.sleep((attempt + 1) * 1000);
      }
    }
    throw new Error('Download failed after retries');
  }

  /**
   * Internal implementation of file download
   */
  private async downloadFileInternal(
    fileId: string,
    offset: number,
    length: number
  ): Promise<Buffer> {
    const [groupName, remoteFilename] = splitFileId(fileId);

    // Get storage server for download
    const storageServer = await this.getDownloadStorageServer(groupName, remoteFilename);

    // Get connection
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Build request
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');
      const bodyLen = 16 + remoteFilenameBytes.length;
      const header = encodeHeader(bodyLen, StorageCommand.DOWNLOAD_FILE, 0);

      const body = Buffer.concat([
        encodeInt64(offset),
        encodeInt64(length),
        remoteFilenameBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      if (respHeader.length <= 0) {
        return Buffer.alloc(0);
      }

      // Receive file data
      const data = await conn.receiveFull(respHeader.length, this.networkTimeout);
      return data;
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Gets a storage server from tracker for download
   */
  private async getDownloadStorageServer(
    groupName: string,
    remoteFilename: string
  ): Promise<StorageServer> {
    const conn = await this.trackerPool.get();

    try {
      // Build request
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');
      const bodyLen = FDFS_GROUP_NAME_MAX_LEN + remoteFilenameBytes.length;
      const header = encodeHeader(bodyLen, TrackerCommand.SERVICE_QUERY_FETCH_ONE, 0);

      const body = Buffer.concat([
        padString(groupName, FDFS_GROUP_NAME_MAX_LEN),
        remoteFilenameBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      const respBody = await conn.receiveFull(respHeader.length, this.networkTimeout);

      // Parse response
      if (respBody.length < FDFS_GROUP_NAME_MAX_LEN + IP_ADDRESS_SIZE + 8) {
        throw new InvalidResponseError('Download storage server response too short');
      }

      let offset = FDFS_GROUP_NAME_MAX_LEN;
      const ipAddr = unpadString(respBody.slice(offset, offset + IP_ADDRESS_SIZE));
      offset += IP_ADDRESS_SIZE;

      const port = decodeInt64(respBody.slice(offset, offset + 8));

      return { ipAddr, port, storePathIndex: 0 };
    } finally {
      this.trackerPool.put(conn);
    }
  }

  /**
   * Downloads a file and saves it to the local filesystem
   */
  async downloadToFile(fileId: string, localFilename: string): Promise<void> {
    const data = await this.downloadFile(fileId, 0, 0);
    writeFileContent(localFilename, data);
  }

  /**
   * Deletes a file from FastDFS
   */
  async deleteFile(fileId: string): Promise<void> {
    for (let attempt = 0; attempt < this.retryCount; attempt++) {
      try {
        await this.deleteFileInternal(fileId);
        return;
      } catch (error) {
        if (attempt === this.retryCount - 1) {
          throw error;
        }
        await this.sleep((attempt + 1) * 1000);
      }
    }
  }

  /**
   * Internal implementation of file deletion
   */
  private async deleteFileInternal(fileId: string): Promise<void> {
    const [groupName, remoteFilename] = splitFileId(fileId);

    // Get storage server
    const storageServer = await this.getDownloadStorageServer(groupName, remoteFilename);

    // Get connection
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Build request
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');
      const bodyLen = FDFS_GROUP_NAME_MAX_LEN + remoteFilenameBytes.length;
      const header = encodeHeader(bodyLen, StorageCommand.DELETE_FILE, 0);

      const body = Buffer.concat([
        padString(groupName, FDFS_GROUP_NAME_MAX_LEN),
        remoteFilenameBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Sets metadata for a file
   */
  async setMetadata(fileId: string, metadata: Metadata, flag: MetadataFlag): Promise<void> {
    const [groupName, remoteFilename] = splitFileId(fileId);

    // Get storage server
    const storageServer = await this.getDownloadStorageServer(groupName, remoteFilename);

    // Get connection
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Encode metadata
      const metadataBytes = encodeMetadata(metadata);
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');

      // Build request
      const bodyLen =
        2 * 8 + 1 + FDFS_GROUP_NAME_MAX_LEN + remoteFilenameBytes.length + metadataBytes.length;
      const header = encodeHeader(bodyLen, StorageCommand.SET_METADATA, 0);

      const body = Buffer.concat([
        encodeInt64(remoteFilenameBytes.length),
        encodeInt64(metadataBytes.length),
        Buffer.from([flag]),
        padString(groupName, FDFS_GROUP_NAME_MAX_LEN),
        remoteFilenameBytes,
        metadataBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Retrieves metadata for a file
   */
  async getMetadata(fileId: string): Promise<Metadata> {
    const [groupName, remoteFilename] = splitFileId(fileId);

    // Get storage server
    const storageServer = await this.getDownloadStorageServer(groupName, remoteFilename);

    // Get connection
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Build request
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');
      const bodyLen = FDFS_GROUP_NAME_MAX_LEN + remoteFilenameBytes.length;
      const header = encodeHeader(bodyLen, StorageCommand.GET_METADATA, 0);

      const body = Buffer.concat([
        padString(groupName, FDFS_GROUP_NAME_MAX_LEN),
        remoteFilenameBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      if (respHeader.length <= 0) {
        return {};
      }

      const respBody = await conn.receiveFull(respHeader.length, this.networkTimeout);

      // Decode metadata
      return decodeMetadata(respBody);
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Retrieves file information
   */
  async getFileInfo(fileId: string): Promise<FileInfo> {
    const [groupName, remoteFilename] = splitFileId(fileId);

    // Get storage server
    const storageServer = await this.getDownloadStorageServer(groupName, remoteFilename);

    // Get connection
    const storageAddr = `${storageServer.ipAddr}:${storageServer.port}`;
    this.storagePool.addAddr(storageAddr);
    const conn = await this.storagePool.get(storageAddr);

    try {
      // Build request
      const remoteFilenameBytes = Buffer.from(remoteFilename, 'utf8');
      const bodyLen = FDFS_GROUP_NAME_MAX_LEN + remoteFilenameBytes.length;
      const header = encodeHeader(bodyLen, StorageCommand.QUERY_FILE_INFO, 0);

      const body = Buffer.concat([
        padString(groupName, FDFS_GROUP_NAME_MAX_LEN),
        remoteFilenameBytes,
      ]);

      // Send request
      await conn.send(header, this.networkTimeout);
      await conn.send(body, this.networkTimeout);

      // Receive response
      const respHeaderData = await conn.receiveFull(10, this.networkTimeout);
      const respHeader = decodeHeader(respHeaderData);

      if (respHeader.status !== 0) {
        const error = mapStatusToError(respHeader.status);
        if (error) throw error;
      }

      const respBody = await conn.receiveFull(respHeader.length, this.networkTimeout);

      // Parse file info (file_size, create_time, crc32, source_ip)
      if (respBody.length < 8 + 8 + 4 + IP_ADDRESS_SIZE) {
        throw new InvalidResponseError('File info response too short');
      }

      const fileSize = decodeInt64(respBody.slice(0, 8));
      const createTimestamp = decodeInt64(respBody.slice(8, 16));
      const crc32 = decodeInt32(respBody.slice(16, 20));
      const sourceIp = unpadString(respBody.slice(20, 20 + IP_ADDRESS_SIZE));

      const createTime = new Date(createTimestamp * 1000);

      return {
        fileSize,
        createTime,
        crc32,
        sourceIpAddr: sourceIp,
      };
    } finally {
      this.storagePool.put(conn);
    }
  }

  /**
   * Sleep utility for retry delays
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: ruby_client/lib/fastdfs/operations.rb
# FastDFS Operations Implementation
#
# This module implements all file operations for the FastDFS client,
# including upload, download, delete, metadata management, and appender
# file operations.
#
# All operations are implemented with retry logic, error handling, and
# proper protocol communication with tracker and storage servers.
#
# # Copyright (C) 2025 FastDFS Ruby Client Contributors
#
# FastDFS may be copied only under the terms of the GNU General
# Public License V3, which may be found in the FastDFS source kit.

require 'fileutils'
require_relative 'connection_pool'
require_relative 'protocol'
require_relative 'types'
require_relative 'errors'

module FastDFS
  # Operations handler for FastDFS file operations.
  #
  # This class handles all file operations including upload, download, delete,
  # metadata management, and appender file operations. It uses connection pools
  # for tracker and storage servers and implements retry logic for failed operations.
  #
  # Operations are not called directly by client code; they are used internally
  # by the Client class.
  class Operations
    # Initializes a new Operations handler.
    #
    # This constructor creates an operations handler that will use the provided
    # connection pools and configuration for all file operations.
    #
    # @param tracker_pool [ConnectionPool] Connection pool for tracker servers.
    # @param storage_pool [ConnectionPool] Connection pool for storage servers.
    # @param network_timeout [Float] Network I/O timeout in seconds.
    # @param retry_count [Integer] Number of retries for failed operations.
    def initialize(tracker_pool:, storage_pool:, network_timeout:, retry_count:)
      # Store connection pools
      # These are used for all network operations
      @tracker_pool = tracker_pool
      @storage_pool = storage_pool
      
      # Store network timeout
      # Used for all socket I/O operations
      @network_timeout = network_timeout
      
      # Store retry count
      # Used for retry logic on failed operations
      @retry_count = retry_count
      
      # Operations handler initialized
      # Ready to perform file operations
    end
    
    # Uploads a file from the local filesystem.
    #
    # This method reads the file from the local filesystem and uploads it to
    # a storage server. It handles retry logic, error handling, and protocol
    # communication.
    #
    # @param local_filename [String] Path to the local file to upload.
    # @param metadata [Hash<String, String>, nil] Optional metadata key-value pairs.
    # @param is_appender [Boolean] Whether this is an appender file (default: false).
    #
    # @return [String] The file ID for the uploaded file.
    #
    # @raise [FileNotFoundError] If the local file does not exist.
    # @raise [NetworkError] If network communication fails.
    # @raise [StorageError] If the storage server reports an error.
    def upload_file(local_filename, metadata, is_appender: false)
      # Validate local filename
      # Must not be nil or empty
      raise InvalidArgumentError.new("local_filename cannot be nil") if local_filename.nil?
      
      # Check if file exists
      # Must exist before we can upload it
      unless File.exist?(local_filename)
        raise FileNotFoundError.new("local file not found: #{local_filename}")
      end
      
      # Read file content
      # Must read in binary mode for binary files
      begin
        file_data = File.read(local_filename, mode: 'rb')
      rescue => e
        raise FileNotFoundError.new("failed to read file: #{local_filename} - #{e.message}")
      end
      
      # Get file extension
      # Extract extension from filename
      ext_name = File.extname(local_filename)
      
      # Remove leading dot
      # Protocol expects extension without dot
      ext_name = ext_name[1..-1] if ext_name.start_with?('.')
      
      # Use default extension if empty
      # Some files may not have extensions
      ext_name = 'bin' if ext_name.empty?
      
      # Upload buffer with file data
      # Delegate to upload_buffer method
      upload_buffer(file_data, ext_name, metadata, is_appender: is_appender)
    end
    
    # Uploads data from a byte buffer.
    #
    # This method uploads raw binary data directly to FastDFS without requiring
    # a file on the local filesystem. It handles retry logic, error handling,
    # and protocol communication.
    #
    # @param data [String] The file content as binary data.
    # @param file_ext_name [String] File extension without dot (e.g., "jpg", "txt").
    # @param metadata [Hash<String, String>, nil] Optional metadata key-value pairs.
    # @param is_appender [Boolean] Whether this is an appender file (default: false).
    #
    # @return [String] The file ID for the uploaded file.
    #
    # @raise [InvalidArgumentError] If parameters are invalid.
    # @raise [NetworkError] If network communication fails.
    # @raise [StorageError] If the storage server reports an error.
    def upload_buffer(data, file_ext_name, metadata, is_appender: false)
      # Validate data
      # Must not be nil or empty
      raise InvalidArgumentError.new("data cannot be nil") if data.nil?
      raise InvalidArgumentError.new("data cannot be empty") if data.empty?
      
      # Validate file extension
      # Must not be nil or empty
      raise InvalidArgumentError.new("file_ext_name cannot be nil") if file_ext_name.nil?
      raise InvalidArgumentError.new("file_ext_name cannot be empty") if file_ext_name.empty?
      
      # Validate extension length
      # Protocol limits extension to 6 characters
      if file_ext_name.bytesize > FDFS_FILE_EXT_NAME_MAX_LEN
        raise InvalidArgumentError.new("file_ext_name too long: maximum #{FDFS_FILE_EXT_NAME_MAX_LEN} bytes")
      end
      
      # Perform upload with retry logic
      # Retry on transient failures
      last_error = nil
      @retry_count.times do |attempt|
        begin
          # Attempt upload
          # Try to upload the file
          file_id = _upload_buffer_internal(data, file_ext_name, metadata, is_appender)
          
          # Upload successful
          # Return file ID
          return file_id
        rescue NetworkError, ConnectionTimeoutError, StorageServerOfflineError => e
          # Transient error, retry
          # Store error for final raise if all retries fail
          last_error = e
          
          # Don't retry on last attempt
          # Will raise error after loop
          next if attempt < @retry_count - 1
        rescue => e
          # Non-transient error, don't retry
          # Raise immediately
          raise e
        end
        
        # Wait before retry
        # Exponential backoff
        sleep(0.1 * (2 ** attempt)) if attempt < @retry_count - 1
      end
      
      # All retries failed
      # Raise last error
      raise last_error
    end
    
    # Downloads a file from FastDFS.
    #
    # This method downloads file content from a storage server. It handles
    # retry logic, error handling, and protocol communication. Supports
    # partial downloads via offset and length parameters.
    #
    # @param file_id [String] The file ID to download.
    # @param offset [Integer] Starting byte offset (default: 0).
    # @param length [Integer] Number of bytes to download (0 means to end, default: 0).
    #
    # @return [String] The file content as binary data.
    #
    # @raise [InvalidFileIDError] If the file ID format is invalid.
    # @raise [FileNotFoundError] If the file does not exist.
    # @raise [NetworkError] If network communication fails.
    # @raise [StorageError] If the storage server reports an error.
    def download_file(file_id, offset: 0, length: 0)
      # Validate file ID
      # Must not be nil or empty
      raise InvalidArgumentError.new("file_id cannot be nil") if file_id.nil?
      raise InvalidArgumentError.new("file_id cannot be empty") if file_id.empty?
      
      # Validate offset
      # Must be non-negative
      raise InvalidArgumentError.new("offset must be >= 0") if offset < 0
      
      # Validate length
      # Must be non-negative
      raise InvalidArgumentError.new("length must be >= 0") if length < 0
      
      # Perform download with retry logic
      # Retry on transient failures
      last_error = nil
      @retry_count.times do |attempt|
        begin
          # Attempt download
          # Try to download the file
          data = _download_file_internal(file_id, offset, length)
          
          # Download successful
          # Return file data
          return data
        rescue NetworkError, ConnectionTimeoutError, StorageServerOfflineError => e
          # Transient error, retry
          # Store error for final raise if all retries fail
          last_error = e
          
          # Don't retry on last attempt
          # Will raise error after loop
          next if attempt < @retry_count - 1
        rescue => e
          # Non-transient error, don't retry
          # Raise immediately
          raise e
        end
        
        # Wait before retry
        # Exponential backoff
        sleep(0.1 * (2 ** attempt)) if attempt < @retry_count - 1
      end
      
      # All retries failed
      # Raise last error
      raise last_error
    end
    
    # Downloads a file and saves it to the local filesystem.
    #
    # This method downloads file content and writes it to a local file in one
    # operation. It handles retry logic, error handling, and file I/O.
    #
    # @param file_id [String] The file ID to download.
    # @param local_filename [String] Path where to save the downloaded file.
    #
    # @raise [InvalidFileIDError] If the file ID format is invalid.
    # @raise [FileNotFoundError] If the file does not exist.
    # @raise [IOError] If unable to write to the local file.
    # @raise [NetworkError] If network communication fails.
    def download_to_file(file_id, local_filename)
      # Validate local filename
      # Must not be nil or empty
      raise InvalidArgumentError.new("local_filename cannot be nil") if local_filename.nil?
      raise InvalidArgumentError.new("local_filename cannot be empty") if local_filename.empty?
      
      # Download file content
      # Get file data from server
      data = download_file(file_id, offset: 0, length: 0)
      
      # Write to local file
      # Must write in binary mode
      begin
        # Create directory if needed
        # Parent directory must exist
        dir = File.dirname(local_filename)
        FileUtils.mkdir_p(dir) unless dir == '.' || dir == ''
        
        # Write file content
        # Open in binary write mode
        File.open(local_filename, 'wb') do |f|
          f.write(data)
        end
      rescue => e
        raise IOError.new("failed to write file: #{local_filename} - #{e.message}")
      end
    end
    
    # Deletes a file from FastDFS.
    #
    # This method sends a delete command to the storage server. It handles
    # retry logic, error handling, and protocol communication.
    #
    # @param file_id [String] The file ID to delete.
    #
    # @raise [InvalidFileIDError] If the file ID format is invalid.
    # @raise [FileNotFoundError] If the file does not exist.
    # @raise [NetworkError] If network communication fails.
    # @raise [StorageError] If the storage server reports an error.
    def delete_file(file_id)
      # Validate file ID
      # Must not be nil or empty
      raise InvalidArgumentError.new("file_id cannot be nil") if file_id.nil?
      raise InvalidArgumentError.new("file_id cannot be empty") if file_id.empty?
      
      # Perform delete with retry logic
      # Retry on transient failures
      last_error = nil
      @retry_count.times do |attempt|
        begin
          # Attempt delete
          # Try to delete the file
          _delete_file_internal(file_id)
          
          # Delete successful
          # Return without error
          return
        rescue NetworkError, ConnectionTimeoutError, StorageServerOfflineError => e
          # Transient error, retry
          # Store error for final raise if all retries fail
          last_error = e
          
          # Don't retry on last attempt
          # Will raise error after loop
          next if attempt < @retry_count - 1
        rescue => e
          # Non-transient error, don't retry
          # Raise immediately
          raise e
        end
        
        # Wait before retry
        # Exponential backoff
        sleep(0.1 * (2 ** attempt)) if attempt < @retry_count - 1
      end
      
      # All retries failed
      # Raise last error
      raise last_error
    end
    
    # Placeholder methods for other operations
    # These would be implemented similarly with retry logic and error handling
    
    # Uploads a slave file (placeholder - not fully implemented)
    def upload_slave_file(master_file_id, prefix_name, file_ext_name, data, metadata)
      # Placeholder implementation
      # This would implement slave file upload logic
      raise OperationNotSupportedError.new("slave file upload not yet implemented")
    end
    
    # Appends data to an appender file (placeholder)
    def append_file(file_id, data)
      # Placeholder implementation
      # This would implement append operation
      raise OperationNotSupportedError.new("append file not yet implemented")
    end
    
    # Modifies content of an appender file (placeholder)
    def modify_file(file_id, offset, data)
      # Placeholder implementation
      # This would implement modify operation
      raise OperationNotSupportedError.new("modify file not yet implemented")
    end
    
    # Truncates an appender file (placeholder)
    def truncate_file(file_id, size)
      # Placeholder implementation
      # This would implement truncate operation
      raise OperationNotSupportedError.new("truncate file not yet implemented")
    end
    
    # Sets metadata for a file (placeholder)
    def set_metadata(file_id, metadata, flag)
      # Placeholder implementation
      # This would implement set metadata operation
      raise OperationNotSupportedError.new("set metadata not yet implemented")
    end
    
    # Gets metadata for a file (placeholder)
    def get_metadata(file_id)
      # Placeholder implementation
      # This would implement get metadata operation
      raise OperationNotSupportedError.new("get metadata not yet implemented")
    end
    
    # Gets file information (placeholder)
    def get_file_info(file_id)
      # Placeholder implementation
      # This would implement get file info operation
      raise OperationNotSupportedError.new("get file info not yet implemented")
    end
    
    private
    
    # Internal method to upload buffer (implementation placeholder)
    def _upload_buffer_internal(data, file_ext_name, metadata, is_appender)
      # Placeholder implementation
      # This would implement the actual upload protocol
      raise OperationNotSupportedError.new("upload not yet fully implemented")
    end
    
    # Internal method to download file (implementation placeholder)
    def _download_file_internal(file_id, offset, length)
      # Placeholder implementation
      # This would implement the actual download protocol
      raise OperationNotSupportedError.new("download not yet fully implemented")
    end
    
    # Internal method to delete file (implementation placeholder)
    def _delete_file_internal(file_id)
      # Placeholder implementation
      # This would implement the actual delete protocol
      raise OperationNotSupportedError.new("delete not yet fully implemented")
    end
  end
end

